core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -2868,35 +2876,36 @@ class KafkaApis(val requestChannel: RequestChannel,
     }
   }
 
-  private def authorize(request: RequestChannel.Request,
-                        operation: AclOperation,
-                        resourceType: ResourceType,
-                        resourceName: String,
-                        logIfAllowed: Boolean = true,
-                        logIfDenied: Boolean = true,
-                        refCount: Int = 1): Boolean = {
+  // private package for testing
+  private[server] def authorize(requestContext: RequestContext,
+                                operation: AclOperation,
+                                resourceType: ResourceType,
+                                resourceName: String,
+                                logIfAllowed: Boolean = true,
+                                logIfDenied: Boolean = true,
+                                refCount: Int = 1): Boolean = {
     authorizer.forall { authZ =>
       val resource = new ResourcePattern(resourceType, resourceName, PatternType.LITERAL)
       val actions = Collections.singletonList(new Action(operation, resource, refCount, logIfAllowed, logIfDenied))
-      authZ.authorize(request.context, actions).asScala.head == AuthorizationResult.ALLOWED
+      authZ.authorize(requestContext, actions).asScala.head == AuthorizationResult.ALLOWED
     }
   }
 
-  private def filterAuthorized(request: RequestChannel.Request,
-                               operation: AclOperation,
-                               resourceType: ResourceType,
-                               resourceNames: Seq[String],
-                               logIfAllowed: Boolean = true,
-                               logIfDenied: Boolean = true): Set[String] = {
+  // private package for testing
+  private[server] def filterAuthorized(requestContext: RequestContext,
+                                       operation: AclOperation,
+                                       resourceType: ResourceType,
+                                       resourceNames: Seq[String],
+                                       logIfAllowed: Boolean = true,
+                                       logIfDenied: Boolean = true): Set[String] = {
     authorizer match {
       case Some(authZ) =>
         val groupedResourceNames = resourceNames.groupBy(identity)
-        val actions = resourceNames.map { resourceName =>
-          val count = groupedResourceNames(resourceName).size
+        val actions = groupedResourceNames.map { case (resourceName, groupedNames) =>
```

Comment:
The ordering is relevant (see the `zip` call below) and we lose it with this change. Did the tests pass with this change? If so, maybe we need to add more tests as they should not.

Seems like the easiest fix may be to call `resourceNames.distinct` before using it anywhere else.

Reply:
Oh, right. Let me fix that.

Reply:
I have extended the unit test to cover this case and updated the implementation as well.

Reply:
To verify, the test case failed without the second fix?